extensions/interactions/NumericExpressionInput/directives/oppia-interactive-numeric-expression-input.component.ts

```diff
@@ -72,19 +73,31 @@ angular.module('oppia').component('oppiaInteractiveNumericExpressionInput', {
           ctrl.value, NumericExpressionInputRulesService);
       };
 
+      ctrl.showOSK = function() {
+        GuppyInitializationService.setShowOSK(true);
+      };
+
       ctrl.$onInit = function() {
         ctrl.hasBeenTouched = false;
         GuppyConfigurationService.init();
         GuppyInitializationService.init('guppy-div-learner');
-        Guppy.event('change', () => {
-          let activeGuppyObject = (
+        let eventType = (
+          DeviceInfoService.isMobileUserAgent() &&
+          DeviceInfoService.hasTouchEvents()) ? 'focus' : 'change';
+        // We need the 'focus' event while using the on screen keyboard (only
+        // for touch-based devices) to capture input from user and the 'change'
+        // event while using the normal keyboard.
+        Guppy.event(eventType, () => {
+          var activeGuppyObject = (
             GuppyInitializationService.findActiveGuppyObject());
           if (activeGuppyObject !== undefined) {
             ctrl.hasBeenTouched = true;
             ctrl.value = activeGuppyObject.guppyInstance.asciimath();
-            // Need to manually trigger the digest cycle to make any 'watchers'
-            // aware of changes in answer.
-            $scope.$apply();
+            if (eventType === 'change') {
+              // Need to manually trigger the digest cycle to make any
+              // 'watchers' aware of changes in answer.
+              $scope.$apply();
+            }
```

Comment:
This code is copied into three places, should it be extracted to some helper function?

Reply:
Actually a lot of code is common between these interactions and I've tried to refactor it as much as possible by using the GuppyInitializationService and GuppyConfigurationService. But these events need to be added in the components itself since they involve triggering the digest cycle with `$scope.$apply();` so I can't put it in a function in a separate service and call it.